### PR TITLE
fix(session): recover from JSON corruption in session state files

### LINKED ISSUE #3277: [Bug] Session state JSON corruption causes persistent 422 errors (Extra data)
## Bug Description

Session state JSON files become corrupted due to concurrent write race conditions, causing `JSONDecodeError: Extra data` on every subsequent request to the affected session. Once corrupted, the session is permanently broken (all messages return 422) until the file is manually repaired.

## Reproduction

Running CoPaw with QQ channel over several hours. The session file grows to ~400KB+. Concurrent `save_session_state` / `update_session_state` calls interleave their writes, producing a file with a valid JSON object followed by garbage fragments.

Error log:
```
ERROR copaw/app/runner/runner.py:564 | Error in query handler: [422] AGENT_UNKNOWN_ERROR: 
Unknown agent error: JSONDecodeError: Extra data: line 1 column 321545 (char 321544)
```

Corrupted file analysis:
```python
File size: 321561 chars
First JSON object ends at char 321544
Remaining chars: 17
Remaining content: '==============\"}}'
```

## Root Cause

`SafeJSONSession` has three file operations (`save_session_state`, `update_session_state`, `get_session_state_dict`) accessing the same session file without any locking. `save_session_state` uses synchronous `open("w")` while `load_session_state`/`update_session_state` use `aiofiles`. When two coroutines execute `open("w")` nearly simultaneously, OS-level writes interleave, producing concatenated JSON fragments.

## Proposed Fix

Add `JSONDecodeError` catch with `raw_decode` fallback at all three `json.loads` call sites in `SafeJSONSession`. When corruption is detected, the first valid JSON object is extracted and a warning is logged, instead of crashing.

## Environment
- CoPaw version: 1.0.2
- Channel: QQ
- Python: 3.13
- OS: Ubuntu 25.10

### PR DESCRIPTION
**Pull request by zealonexp**
## Description

Fix a P0 availability issue where session state JSON files become corrupted due to concurrent write race conditions, causing `JSONDecodeError: Extra data` on every subsequent request to the affected session.

**Root cause:** `save_session_state`, `update_session_state`, and `get_session_state_dict` access the same session file concurrently without any locking mechanism. When two coroutines execute `open("w")` on the same file nearly simultaneously, the OS-level writes interleave, producing a file with a valid JSON object followed by garbage fragments. Once corrupted, every request to that session fails with a 422 error — no recovery is possible without manual intervention.

**Fix:** Add `JSONDecodeError` catch with `raw_decode` fallback at all three `json.loads` call sites in `SafeJSONSession`. When corruption is detected, the first valid JSON object is extracted and a warning is logged, instead of crashing the request.

**Related Issue:** Fixes #3277

**Security Considerations:** No changes to auth, config handling, or external-facing interfaces. Pure defensive parsing change.

## Type of Change

- [x] Bug fix

## Component(s) Affected

- [x] Core / Backend (app, agents, config, providers, utils, local_models)

## Checklist

- [x] I ran `pre-commit run --all-files` locally and it passes
- [x] If pre-commit auto-fixed files, I committed those changes and reran checks
- [x] I ran tests locally (`pytest` or as relevant) and they pass
- [x] Documentation updated (if needed)
- [x] Ready for review

### New tests added

`tests/unit/agents/test_session.py` — 7 test cases covering:

| Test | Scenario |
|------|----------|
| `test_load_valid_json` | Normal JSON loads without error (baseline) |
| `test_load_corrupted_json_extra_data` | 17-char garbage tail (first observed pattern) |
| `test_load_corrupted_json_real_world_tail` | 203-char garbage tail (second observed pattern) |
| `test_get_corrupted_json` | `get_session_state_dict` recovers from corruption |
| `test_update_corrupted_json` | `update_session_state` recovers, then writes clean JSON |
| `test_load_nonexistent` | Non-existent session returns gracefully |
| `test_get_nonexistent` | Non-existent session returns empty dict |

## Local Verification Evidence

```bash
pre-commit run --files src/copaw/app/runner/session.py tests/unit/agents/test_session.py
# check python ast.........................................Passed
# check docstring is first.........................................Passed
# detect private key.......................................................Passed
# trim trailing whitespace.................................................Passed
# Add trailing commas......................................................Passed
# mypy.....................................................................Passed
# black....................................................................Passed
# flake8...................................................................Passed
# pylint...................................................................Passed

pytest tests/unit/agents/test_session.py -v
# 7 passed, 1 warning in 0.16s
```

### Files changed

| File | Change |
|------|--------|
| `src/copaw/app/runner/session.py` | Add `JSONDecodeError` → `raw_decode` fallback in `load_session_state`, `update_session_state`, `get_session_state_dict` |
| `tests/unit/agents/test_session.py` | **New** — 7 test cases for corruption resilience |

### Recommended follow-ups (out of scope)

1. **`asyncio.Lock`** — Serialize concurrent reads/writes to eliminate race at source
2. **Atomic writes** — Write to temp file then `os.rename()`
3. **Session file rotation** — Files grow unboundedly; compaction would help

### CODE REVIEW

**Comment by zealonexp** [comment]
## Update: Branch refreshed + confirmed bug still exists on latest main

Just pushed the latest commit with expanded test coverage (16 test cases total).

### Verified: Bug is still present on upstream main (v1.1.1b1, commit c4ea882)

I diffed `src/qwenpaw/app/runner/session.py` between our PR base and the latest upstream main — **byte-identical**. All three `json.loads()` call sites remain unprotected:

| Method | Risk |
|---|---|
| `load_session_state` | `json.loads(content)` — crashes on corruption |
| `update_session_state` | `json.loads(content)` — crashes on corruption |
| `get_session_state_dict` | `json.loads(content)` — crashes on corruption |

The concurrent write race condition is still reproducible: `save_session_state` and `update_session_state` both use synchronous `open(w)` without any locking, while the read methods use `aiofiles`. When two coroutines write simultaneously, the OS-level writes interleave, producing `JSONDecodeError: Extra data` — every subsequent request to that session fails with HTTP 422.

### Our fix
`_safe_json_loads()` catches `JSONDecodeError` and falls back to `json.JSONDecoder().raw_decode()` to recover the first valid JSON object, logging a warning instead of crashing.

### Test coverage
16 test cases covering:
- Valid JSON (baseline)
- 17-char garbage tail (first observed pattern)
- 203-char garbage tail (second observed pattern)
- Empty file / Multiple concatenated JSON objects / Unicode in corrupted tail
- Plus 10 more edge cases

All tests pass against the latest upstream main. Ready for review.

**Comment by zealonexp** [comment]
## 🔁 Bug still present in v1.1.1-beta.1

I just verified that this concurrency bug still exists in the latest release **v1.1.1-beta.1** (released 2026-04-13).

### What I checked

Cloned `v1.1.1-beta.1` and inspected `src/qwenpaw/app/runner/session.py`:

- **`save_session_state`** (line 85): uses bare `open("w")` — no lock, no atomic write
- **`update_session_state`** (line 181): same bare `open("w")` — no lock, no atomic write
- **`get_session_state_dict`** (line 218): reads without any coordination with concurrent writers

All three methods operate on the **same session JSON file** with zero concurrency protection. When two async coroutines call `save_session_state` and `update_session_state` concurrently on the same session, their `open("w")` calls interleave at the OS level, producing a file like:

```
{"conversation": [...]}{"
```

which causes `json.JSONDecodeError: Extra data` on every subsequent read.

### Reproduction scenario

This is easily triggered in production when:
1. A user sends a message (triggers `save_session_state`)
2. Simultaneously a cron/background task calls `update_session_state` on the same session
3. The writes interleave → file corrupted → all future requests to that session fail

### Impact

- **Severity**: P0 — once corrupted, the session is permanently broken
- **Blast radius**: every request to the affected session returns 500
- **Recovery**: requires manual file deletion or repair

### Recommended fix

At minimum, add file-level locking (e.g., `fcntl.flock` on Linux / `msvcrt.locking` on Windows) around all read-modify-write cycles. For a more robust approach, consider atomic writes (write to temp file + `os.rename`).

Happy to help test any fix — I have been running with the patch from this PR in production for a few days now with zero corruption incidents.

**Comment by xieyxclack** [comment]
Please resolve the conflict first

File: src/copaw/app/runner/session.py [hunk 1]
```diff
@@ -111,7 +111,18 @@ async def load_session_state(
...
+                except json.JSONDecodeError:
+                    # Fallback: extract first valid JSON object (handles
+                    # race-condition corruption where two writes overlap)
+                    decoder = json.JSONDecoder()
+                    states, _ = decoder.raw_decode(content)
+                    logger.warning(
+                        "Session file %s had corrupted JSON (Extra data). "
+                        "Recovered first valid object.",
+                        session_save_path,
+                    )
```
[the remark this answers is not in the document]
Comment (zealonexp) [reply src/copaw/app/runner/session.py]: Already addressed in commit 0bb16b4 (`refactor(session): extract _safe_json_loads to DRY corruption recovery`). The duplicated JSON recovery logic has been extracted into a single `_safe_json_loads()` private method. A module-level `JSONDecoder()` instance is also created once for reuse.

File: src/copaw/app/runner/session.py [hunk 2]
```diff
@@ -154,7 +165,16 @@ async def update_session_state(
                 errors="surrogatepass",
             ) as f:
                 content = await f.read()
-                states = json.loads(content)
+                try:
+                    states = json.loads(content)
+                except json.JSONDecodeError:
+                    decoder = json.JSONDecoder()
+                    states, _ = decoder.raw_decode(content)
```
[the remark this answers is not in the document]
Comment (zealonexp) [reply src/copaw/app/runner/session.py]: Good point. In the current implementation, `_safe_json_loads()` lets `raw_decode` failures propagate as `JSONDecodeError` intentionally — this makes it explicit that the file is beyond recovery (e.g. completely empty or binary garbage). Callers catch this and log appropriately. For `update_session_state` specifically, the `ignore_not_exist` path already falls back gracefully. Adding a blanket empty-dict fallback on total corruption would silently mask data loss, which feels worse than a loud failure. Open to changing this if maintainers prefer otherwise.

**Comment by zealonexp** [comment]
Hi maintainers 👋

Just a heads-up — v1.1.2 shipped today and this session JSON corruption bug is still alive and kicking. The concurrent write race in `SafeJSONSession` can still corrupt session files under load, leaving sessions completely unusable.

This is a P0 issue in production — once a session file gets trampled by concurrent `open("w")` calls, there's no recovery without manual intervention. Would be awesome to see the fix land soon! 🚀

**Comment by xieyxclack** [comment]
@zealonexp Thank you for reporting! We will follow this pr

**Review by xieyxclack — APPROVED** [review]

**Comment by zealonexp** [comment]
> ## Welcome to QwenPaw! 🎉
> Thank you @zealonexp for your first contribution! Your PR has been merged. 🚀
> 
> We'd love to give you a shout-out in our release notes! If you're comfortable sharing, please reply to this comment with your social media handles using the format below:
> 
> ```
> discord: your_discord_handle
> x: your_x_handle
> xiaohongshu: your_xiaohongshu_id
> ```
> 
> > **Note:** Only share what you're comfortable with — all fields are optional.
> 
> Thanks again for helping make QwenPaw better!

discord:Copaw_
x:@Copaw_
